core/services/relay/evm/chain_reader.go

```diff
@@ -63,87 +65,124 @@ func NewChainReaderService(ctx context.Context, lggr logger.Logger, lp logpoller
 			case types.Event:
+				partOfContractCommonFilter := slices.Contains(chainContractReader.GenericEventNames, typeName)
+				if !partOfContractCommonFilter && !chainReaderDefinition.HasPollingFilter() {
+					return fmt.Errorf(
+						"%w: chain reader has no polling filter defined for contract: %s, event: %s",
+						commontypes.ErrInvalidConfig, contractName, typeName)
+				}
+
+				eventOverridesContractFilter := chainReaderDefinition.HasPollingFilter()
+				if eventOverridesContractFilter && partOfContractCommonFilter {
+					return fmt.Errorf(
+						"%w: conflicting chain reader polling filter definitions for contract: %s event: %s, can't have polling filter defined both on contract and event level",
+						commontypes.ErrInvalidConfig, contractName, typeName)
+				}
+
+				if !eventOverridesContractFilter &&
+					!slices.Contains(eventSigsForContractFilter, contractAbi.Events[chainReaderDefinition.ChainSpecificName].ID) {
+					eventSigsForContractFilter = append(eventSigsForContractFilter, contractAbi.Events[chainReaderDefinition.ChainSpecificName].ID)
+				}
+
 				err = cr.addEvent(contractName, typeName, contractAbi, *chainReaderDefinition)
 			default:
 				return fmt.Errorf(
 					"%w: invalid chain reader definition read type: %s",
 					commontypes.ErrInvalidConfig,
 					chainReaderDefinition.ReadType)
 			}
-
 			if err != nil {
 				return err
 			}
 		}
+		cr.contractBindings[contractName].pollingFilter = chainContractReader.PollingFilter.ToLPFilter(eventSigsForContractFilter)
 	}
 	return nil
 }
 
+func (cr *chainReader) Name() string { return cr.lggr.Name() }
+
+// Start registers polling filters if contracts are already bound.
 func (cr *chainReader) Start(ctx context.Context) error {
 	return cr.StartOnce("ChainReader", func() error {
-		return cr.contractBindings.ForEach(ctx, readBinding.Register)
+		return cr.contractBindings.ForEach(ctx, func(c context.Context, cb *contractBinding) error {
+			for _, rb := range cb.readBindings {
+				if err := rb.Register(ctx); err != nil {
+					return err
+				}
+			}
+			return cb.Register(ctx, cr.lp)
+		})
 	})
 }
 
+// Close unregisters polling filters for bound contracts.
 func (cr *chainReader) Close() error {
 	return cr.StopOnce("ChainReader", func() error {
 		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
 		defer cancel()
-		return cr.contractBindings.ForEach(ctx, readBinding.Unregister)
+		return cr.contractBindings.ForEach(ctx, func(c context.Context, cb *contractBinding) error {
+			for _, rb := range cb.readBindings {
+				if err := rb.Unregister(ctx); err != nil {
+					return err
+				}
+			}
+			return cb.Unregister(ctx, cr.lp)
+		})
 	})
 }
 
 func (cr *chainReader) Ready() error { return nil }
+
 func (cr *chainReader) HealthReport() map[string]error {
 	return map[string]error{cr.Name(): nil}
 }
 
+func (cr *chainReader) GetLatestValue(ctx context.Context, contractName, method string, params any, returnVal any) error {
```

Comment:
this needs some documentation. what should the user expect if they pass the wrong type of params and/or return val? how is the user expected to know what the correct types are?

Reply:
There are comments on the interface level, do we usually copy them over to the implementation?

Reply:
I don't think that copying will add value (especially if here or elsewhere already says this is an implementation of that particular interface), but the comment here can elaborate on the implementation details if that is relevant.

Reply:
There aren't any implementation specifics that deserve a comment imho

Reply:
my mistake; i didn't see the comments on the interface.